### ISSUE TITLE
not all datasets get indexed

### ISSUE THREAD

**Issue by smurfio** [arangodb/arangodb#9866]
## My Environment

* __ArangoDB Version__:        3.5.0
* __Storage Engine__:             RocksDB 
* __Deployment Mode__:        Single Server
* __Deployment Strategy__:   Manual Start in Docker
* __Infrastructure__:               GCP
* __Total RAM in your machine__:        32GB
* __Disks in use__:                  SSD
* __Used Package__:              Docker - official Docker library

## Component, Query & Data

__Affected feature__:

## Steps to reproduce

__Problem__:
Since upgrading to 3.5.0, some datasets seem not to be properly indexed. Thus not findable via searching on _from or _to attributes. Unloading and Loading the collection helps. However, this doesn't create a new index on all datasets that are stored within the collection. So some datasets stay out of the index. If I use an upsert statement the system, however, realises that the same data are already existing and throws an error.
I remember that I experienced the same issue with the first edition of arangodb that used Rocksdb as a storage engine.

__Expected result__:
All datasets are within the index.

**Comment by ASolomatin** [comment]
We has totally same issue on 3.5.4 and 3.6.2. For us, this is a show-stopper for bringing arangodb to production.

For example, query
```
FOR a IN 1..1 OUTBOUND "Subjects/44591447" `Founder`
FILTER a._id == "Subjects/2913869"
LET b = FIRST(
    FOR v IN 1..1 INBOUND a `Founder`
    FILTER v._id == "Subjects/44591447"
    RETURN v
)

RETURN {a, b}
```
returns
```
[
  {
    "a": {
      "_key": "2913869",
      "_id": "Subjects/2913869",
      "_rev": "_aMxZrHq--_",
      "_tp": 1,
      "fndc": "13000000.00",
      "_ic": true,
      "cnid": 45949,
      "stts": 1,
      "accd": "46.46"
    },
    "b": null
  }
]
```
but after edge collection `Founder` is reloaded, everything starts work as expected:
```
[
  {
    "a": {
      "_key": "2913869",
      "_id": "Subjects/2913869",
      "_rev": "_aMxZrHq--_",
      "_tp": 1,
      "fndc": "13000000.00",
      "_ic": true,
      "cnid": 45949,
      "stts": 1,
      "accd": "46.46"
    },
    "b": {
      "_key": "44591447",
      "_id": "Subjects/44591447",
      "_rev": "_aM0d9BS---",
      "_tp": 1,
      "fndc": "1000000.00",
      "_ic": true,
      "cnid": 1821608,
      "stts": 1,
      "accd": "86.21"
    }
  }
]
```
Also, how can i see, this bug has very random character: after collection reloaded it stops reproducing on specified documents but later we finding the same on other random documents.

**Comment by danielhlarkin** [comment]
Hey guys, thanks for bringing this to our attention. @smurfio, sorry for the slow response. @ASolomatin, thanks for refreshing the issue. We're hoping to get a better understanding of what's going on and figure out a way to reproduce it. From what I can tell, this happens even in a read-only workload (no writes after reloading the collection). Is that accurate?

Any more details you can share about your environment (platform, CPU, RAM, etc.) would help, as would any more details you can share about your dataset and workload. If you are able to reproduce with a dataset and workload you can share with us, that would be even better.

One particular piece of information that might help would be the engine stats which report cache usage information. You can retrieve them in single-server mode via a `db._engineStats()` call in arangosh or a `GET` to `/_api/engine/stats`.

**Comment by ASolomatin** [comment]
Hi, thanks for quick response.

I don’t remember the cases when this happened in a read-only, but the real testing we started only when we switched to real-time updates (ubdate, delete, create).
Initially database was built in 15 parallell threads that makes a lot of inserts updates and delates while building a compleate graph structure.

Our `Subjects` collection, from the example, today contains 11458996 documents and `Founder` contains 2551495 edges.

About our environment I will tell You tomorow.

Unfortunately, I can’t reproduce this on small test datasets and I can’t share the current dataset because it is based on production data.

Here is the engine stats after reloading 2 collections from 20, later i will be able to share more stats after building new dataset versions.

```
arango362.lcl:8529@RelationsRT2020V5> db._engineStats()
{ 
  "rocksdb.num-files-at-level0" : 4, 
  "rocksdb.compression-ratio-at-level0" : 4, 
  "rocksdb.num-files-at-level1" : 0, 
  "rocksdb.compression-ratio-at-level1" : 0, 
  "rocksdb.num-files-at-level2" : 0, 
  "rocksdb.compression-ratio-at-level2" : 0, 
  "rocksdb.num-files-at-level3" : 0, 
  "rocksdb.compression-ratio-at-level3" : 0, 
  "rocksdb.num-files-at-level4" : 1, 
  "rocksdb.compression-ratio-at-level4" : 1, 
  "rocksdb.num-files-at-level5" : 14, 
  "rocksdb.compression-ratio-at-level5" : 5, 
  "rocksdb.num-files-at-level6" : 122, 
  "rocksdb.compression-ratio-at-level6" : 10, 
  "rocksdb.num-immutable-mem-table" : 0, 
  "rocksdb.num-immutable-mem-table-flushed" : 45, 
  "rocksdb.mem-table-flush-pending" : 0, 
  "rocksdb.compaction-pending" : 0, 
  "rocksdb.background-errors" : 0, 
  "rocksdb.cur-size-active-mem-table" : 147073968, 
  "rocksdb.cur-size-all-mem-tables" : 147073968, 
  "rocksdb.size-all-mem-tables" : 1298580072, 
  "rocksdb.num-entries-active-mem-table" : 9091, 
  "rocksdb.num-entries-imm-mem-tables" : 0, 
  "rocksdb.num-deletes-active-mem-table" : 432, 
  "rocksdb.num-deletes-imm-mem-tables" : 0, 
  "rocksdb.estimate-num-keys" : 259777057, 
  "rocksdb.estimate-table-readers-mem" : 427529500, 
  "rocksdb.num-snapshots" : 0, 
  "rocksdb.oldest-snapshot-time" : 0, 
  "rocksdb.num-live-versions" : 7, 
  "rocksdb.min-log-number-to-keep" : 39024, 
  "rocksdb.estimate-live-data-size" : 7383490159, 
  "rocksdb.live-sst-files-size" : 8128358074, 
  "rocksdb.dbstats" : "\n** DB Stats **\nUptime(secs): 623362.4 total, 556.2 interval\nCumulative writes: 601K writes, 10M keys, 601K commit groups, 1.0 writes per commit group, ingest: 8.96 GB, 0.01 MB/s\nCumulative WAL: 601K writes, 0 syncs, 601399.00 writes per sync, written: 8.9...", 
  "rocksdb.sstables" : "--- level 0 --- version# 838 ---\n--- level 1 --- version# 838 ---\n--- level 2 --- version# 838 ---\n--- level 3 --- version# 838 ---\n--- level 4 --- version# 838 ---\n--- level 5 --- version# 838 ---\n--- level 6 --- version# 838 ---\n 39028:2006185[0 .. 0]['3...", 
  "rocksdb.num-running-compactions" : 0, 
  "rocksdb.num-running-flushes" : 0, 
  "rocksdb.is-file-deletions-enabled" : 1, 
  "rocksdb.estimate-pending-compaction-bytes" : 0, 
  "rocksdb.base-level" : 6, 
  "rocksdb.block-cache-capacity" : 14183018496, 
  "rocksdb.block-cache-usage" : 14181342000, 
  "rocksdb.block-cache-pinned-usage" : 0, 
  "rocksdb.total-sst-files-size" : 8128358074, 
  "rocksdb.actual-delayed-write-rate" : 0, 
  "rocksdb.is-write-stopped" : 0, 
  "cache.limit" : 11819182080, 
  "cache.allocated" : 6573491139, 
  "cache.hit-rate-lifetime" : 55.50302410508559, 
  "cache.hit-rate-recent" : 84.844970703125, 
  "columnFamilies" : { 
    "definitions" : { 
      "dbstats" : "\n** Compaction Stats [default] **\nLevel    Files   Size     Score Read(GB)  Rn(GB) Rnp1(GB) Write(GB) Wnew(GB) Moved(GB) W-Amp Rd(MB/s) Wr(MB/s) Comp(sec) CompMergeCPU(sec) Comp(cnt) Avg(sec) KeyIn KeyDrop\n--------------------------------------------------...", 
      "memory" : 59862185 
    }, 
    "documents" : { 
      "dbstats" : "\n** Compaction Stats [Documents] **\nLevel    Files   Size     Score Read(GB)  Rn(GB) Rnp1(GB) Write(GB) Wnew(GB) Moved(GB) W-Amp Rd(MB/s) Wr(MB/s) Comp(sec) CompMergeCPU(sec) Comp(cnt) Avg(sec) KeyIn KeyDrop\n------------------------------------------------...", 
      "memory" : 4194768573 
    }, 
    "primary" : { 
      "dbstats" : "\n** Compaction Stats [PrimaryIndex] **\nLevel    Files   Size     Score Read(GB)  Rn(GB) Rnp1(GB) Write(GB) Wnew(GB) Moved(GB) W-Amp Rd(MB/s) Wr(MB/s) Comp(sec) CompMergeCPU(sec) Comp(cnt) Avg(sec) KeyIn KeyDrop\n---------------------------------------------...", 
      "memory" : 1114404584 
    }, 
    "edge" : { 
      "dbstats" : "\n** Compaction Stats [EdgeIndex] **\nLevel    Files   Size     Score Read(GB)  Rn(GB) Rnp1(GB) Write(GB) Wnew(GB) Moved(GB) W-Amp Rd(MB/s) Wr(MB/s) Comp(sec) CompMergeCPU(sec) Comp(cnt) Avg(sec) KeyIn KeyDrop\n------------------------------------------------...", 
      "memory" : 2006442462 
    }, 
    "vpack" : { 
      "dbstats" : "\n** Compaction Stats [VPackIndex] **\nLevel    Files   Size     Score Read(GB)  Rn(GB) Rnp1(GB) Write(GB) Wnew(GB) Moved(GB) W-Amp Rd(MB/s) Wr(MB/s) Comp(sec) CompMergeCPU(sec) Comp(cnt) Avg(sec) KeyIn KeyDrop\n-----------------------------------------------...", 
      "memory" : 811486964 
    }, 
    "geo" : { 
      "dbstats" : "\n** Compaction Stats [GeoIndex] **\nLevel    Files   Size     Score Read(GB)  Rn(GB) Rnp1(GB) Write(GB) Wnew(GB) Moved(GB) W-Amp Rd(MB/s) Wr(MB/s) Comp(sec) CompMergeCPU(sec) Comp(cnt) Avg(sec) KeyIn KeyDrop\n-------------------------------------------------...", 
      "memory" : 0 
    }, 
    "fulltext" : { 
      "dbstats" : "\n** Compaction Stats [FulltextIndex] **\nLevel    Files   Size     Score Read(GB)  Rn(GB) Rnp1(GB) Write(GB) Wnew(GB) Moved(GB) W-Amp Rd(MB/s) Wr(MB/s) Comp(sec) CompMergeCPU(sec) Comp(cnt) Avg(sec) KeyIn KeyDrop\n--------------------------------------------...", 
      "memory" : 0 
    } 
  }, 
  "rocksdbengine.throttle.bps" : 0 
}
```

**Comment by ASolomatin** [comment]
Hi. Here is our environment:
Dell T620/630, 2x Intel Xeon E5-2620v3 2.40GHz, DDR4 - 2133 UDIMM
240 SSD Enterprise Intel
Hypervisor: SmartOS, bhyve
VM: Ubuntu 18.04, 16 threads, 48 GB RAM

**Comment by Simran-B** [comment]
Hi @ASolomatin & @smurfio,
there was a similar issue reported for which a fix will become available in v3.5.6 and v3.6.4:

> * Fixed issue #11303: The result data of filter operation on edge document with 
>  _from or _to field ALWAYS miss some data on version 3.6.2.
>
>   This bug could have led to some edges from a RocksDB edge index not being 
>   returned when a non-traversal edge lookup was done in an AQL query in an inner 
>   FOR loop. Whether or not edges were withheld depended on whether edges were
>   located within the same prefix domain.

PRs: #11536 #11537

**Comment by jsteemann** [comment]
I expect the bug fixes in PRs #11536 and #11537 to help here, too.

**Comment by jsteemann** [comment]
@ASolomatin @smurfio @Simran-B : Issue #11303 was fixed yesterday, and there is a release 3.6.3-1 on our website now that contains the fix. 
Do you mind trying if this solves this problem, too?
Thanks!

**Comment by dothebart** [comment]
A fix for this hase been published, closing.

Please note that [ArangoDB 3.7](https://www.arangodb.com/2020/08/arangodb-3-7-a-big-step-forward-for-multi-model/) is available.
